server/src/main/java/io/crate/expression/scalar/HasSchemaPrivilegeFunction.java

```diff
@@ -0,0 +1,342 @@
+        return user;
+    };
+
+    private static final TriFunction<User, Object, List<Privilege.Type>, Boolean> CHECK_BY_SCHEMA_NAME = (user, schema, privileges) -> {
+        String schemaName = (String) schema;
+        boolean result = false;
+        for (Privilege.Type type: privileges) {
+            // USAGE is allowed for public schemas
+            if (type == Privilege.Type.DQL && PgCatalogTableDefinitions.isPgCatalogOrInformationSchema(schemaName)) {
+                return true;
+            }
+            // Last argument is null as it's not used for Privilege.Clazz.SCHEMA
+            result |= user.hasPrivilege(type, Privilege.Clazz.SCHEMA, schemaName, null);
+        }
+        return result;
+    };
+
+    private static final TriFunction<User, Object, List<Privilege.Type>, Boolean> CHECK_BY_SCHEMA_OID = (user, schema, privileges) -> {
+        Integer schemaOid = (Integer) schema;
+        boolean result = false;
+        for (Privilege.Type type: privileges) {
+            // USAGE is allowed for public schemas
+            if (type == Privilege.Type.DQL && PgCatalogTableDefinitions.isPgCatalogOrInformationSchema(schemaOid)) {
+                return true;
+            }
+            result |= user.hasSchemaPrivilege(type, schemaOid); // Returns true if has any privilege out of 2 possible
+        }
+        return result;
+    };
+
+    /**
+     * @param privilege is a comma separated list.
+     * Valid privileges are 'CREATE' and 'USAGE' which map to DDL and DQL correspondingly.
+     * Case of the privilege string is not significant, and extra whitespace is allowed between privilege names.
+     * Repetition of the valid argument is allowed.
+     *
+     * @throws IllegalArgumentException if privilege contains invalid privilege.
+     */
+    @Nullable
+    private static List<Privilege.Type> privilegesToCheck(String privilege) {
+        List<Privilege.Type> toCheck = new ArrayList<>();
+        // toCheck has at most 2 elements, fine to do contains in loop.
+        String[] privileges = privilege.toLowerCase(Locale.ENGLISH).split(",");
+        for (int i = 0; i < privileges.length; i++) {
+            privileges[i] = privileges[i].trim();
+            if (privileges[i].equals("create")) {
+                if (toCheck.contains(Privilege.Type.DDL) == false) {
+                    toCheck.add(Privilege.Type.DDL);
+                }
+            } else if (privileges[i].equals("usage")) {
+                if (toCheck.contains(Privilege.Type.DQL) == false) {
+                    toCheck.add(Privilege.Type.DQL);
+                }
+            } else {
+                // Same error as PG
+                throw new IllegalArgumentException(String.format(Locale.ENGLISH, "Unrecognized privilege type: %s", privileges[i]));
+            }
+        }
+        return toCheck;
+    }
+
+    public static void register(ScalarFunctionModule module) {
+        var provider = module.binder().getProvider(UserLookup.class);
```

Comment:
What about adding the `UserLookup` to the `NodeContext` which is already passed into `evaluate` instead?
This way we'd avoid using the binder here and would be tight to the injection framework (which we may want to get rid eventually).

Reply:
Done in https://github.com/crate/crate/pull/12553/commits/a702bad81e0ebb6691aed6a3090781dffe7e5a9b
Also added `userLookup` to `compile`